ffi/transforms.cpp

```diff
@@ -1,96 +1,272 @@
+
 API_EXPORT(void)
-LLVMPY_PassManagerBuilderPopulateModulePassManager(
-    LLVMPassManagerBuilderRef PMB, LLVMPassManagerRef PM) {
-    LLVMPassManagerBuilderPopulateModulePassManager(PMB, PM);
+LLVMPY_LLVMLoopAnalysisManagerDispose(LLVMLoopAnalysisManager LAM) {
+    delete (LAM);
 }
 
-API_EXPORT(unsigned)
-LLVMPY_PassManagerBuilderGetOptLevel(LLVMPassManagerBuilderRef PMB) {
-    llvm::PassManagerBuilder *pmb = llvm::unwrap(PMB);
-    return pmb->OptLevel;
+API_EXPORT(LLVMFunctionAnalysisManager)
+LLVMPY_LLVMFunctionAnalysisManagerCreate() {
+    return new llvm::FunctionAnalysisManager;
 }
 
 API_EXPORT(void)
-LLVMPY_PassManagerBuilderSetOptLevel(LLVMPassManagerBuilderRef PMB,
-                                     unsigned OptLevel) {
-    LLVMPassManagerBuilderSetOptLevel(PMB, OptLevel);
+LLVMPY_LLVMFunctionAnalysisManagerDispose(LLVMFunctionAnalysisManager FAM) {
+    delete (FAM);
 }
 
-API_EXPORT(unsigned)
-LLVMPY_PassManagerBuilderGetSizeLevel(LLVMPassManagerBuilderRef PMB) {
-    llvm::PassManagerBuilder *pmb = llvm::unwrap(PMB);
-    return pmb->SizeLevel;
+API_EXPORT(LLVMCGSCCAnalysisManager)
+LLVMPY_LLVMCGSCCAnalysisManagerCreate() {
+    return new llvm::CGSCCAnalysisManager;
 }
 
 API_EXPORT(void)
-LLVMPY_PassManagerBuilderSetSizeLevel(LLVMPassManagerBuilderRef PMB,
-                                      unsigned SizeLevel) {
-    LLVMPassManagerBuilderSetSizeLevel(PMB, SizeLevel);
+LLVMPY_LLVMCGSCCAnalysisManagerDispose(LLVMCGSCCAnalysisManager CGAM) {
+    delete (CGAM);
 }
 
-API_EXPORT(int)
-LLVMPY_PassManagerBuilderGetDisableUnrollLoops(LLVMPassManagerBuilderRef PMB) {
-    llvm::PassManagerBuilder *pmb = llvm::unwrap(PMB);
-    return pmb->DisableUnrollLoops;
+API_EXPORT(LLVMPassInstrumentationCallbacks)
+LLVMPY_LLVMPassInstrumentationCallbacksCreate() {
+    return new llvm::PassInstrumentationCallbacks;
+}
+
+API_EXPORT(void)
+LLVMPY_LLVMPassInstrumentationCallbacksDispose(
+    LLVMPassInstrumentationCallbacks PIC) {
+    delete (PIC);
 }
 
 API_EXPORT(void)
-LLVMPY_PassManagerBuilderSetDisableUnrollLoops(LLVMPassManagerBuilderRef PMB,
-                                               LLVMBool Value) {
-    LLVMPassManagerBuilderSetDisableUnrollLoops(PMB, Value);
+LLVMPY_PassManagerBuilderDispose(LLVMPassBuilder PB,
+                                 LLVMPassBuilderOptionsRef Options) {
+    // TODO figure out proper deletion
+    // delete(PB);
+    // LLVMDisposePassBuilderOptions(Options);
+}
+
+// Deprecated: no longer exists in LLVM
+API_EXPORT(LLVMModulePassManager)
+LLVMPY_PassManagerBuilderPopulateModulePassManager(
+    LLVMPassBuilder PB, LLVMPassBuilderOptionsRef Options,
+    LLVMOptimizationLevel Level, LLVMModulePassManager MPM,
+    LLVMModuleAnalysisManager MAM, LLVMLoopAnalysisManager LAM,
+    LLVMFunctionAnalysisManager FAM, LLVMCGSCCAnalysisManager CGAM,
+    LLVMPassInstrumentationCallbacks PIC) {
+    // TODO handle PB memory better
+    if (PB)
```

Comment:
This is here because the PipelineTuningOptions cannot currently be updated after creation of a PassBuilder. I'm planning on floating the idea to LLVM upstream of allowing this change so we don't have to do this stuff